src/server/game/Globals/ObjectMgr.cpp

```diff
@@ -8702,122 +8689,139 @@ void ObjectMgr::LoadMailLevelRewards()
     TC_LOG_INFO("server.loading", ">> Loaded %u level dependent mail rewards in %u ms", count, GetMSTimeDiffToNow(oldMSTime));
 }
 
-void ObjectMgr::AddSpellToTrainer(uint32 ID, uint32 SpellID, uint32 MoneyCost, uint32 ReqSkillLine, uint32 ReqSkillRank, uint32 ReqLevel)
+void ObjectMgr::LoadTrainers()
 {
-    if (ID >= TRINITY_TRAINER_START_REF)
-        return;
-
-    CreatureTemplate const* cInfo = GetCreatureTemplate(ID);
-    if (!cInfo)
-    {
-        TC_LOG_ERROR("sql.sql", "Table `npc_trainer` contains entries for a non-existing creature template (ID: %u), ignoring", ID);
-        return;
-    }
-
-    if (!(cInfo->npcflag & UNIT_NPC_FLAG_TRAINER))
-    {
-        TC_LOG_ERROR("sql.sql", "Table `npc_trainer` contains entries for a creature template (ID: %u) without trainer flag, ignoring", ID);
-        return;
-    }
-
-    SpellInfo const* spellinfo = sSpellMgr->GetSpellInfo(SpellID);
-    if (!spellinfo)
-    {
-        TC_LOG_ERROR("sql.sql", "Table `npc_trainer` contains an ID (%u) for a non-existing SpellID (%u), ignoring", ID, SpellID);
-        return;
-    }
+    uint32 oldMSTime = getMSTime();
 
-    if (!SpellMgr::IsSpellValid(spellinfo))
-    {
-        TC_LOG_ERROR("sql.sql", "Table `npc_trainer` contains an ID (%u) for a broken SpellID (%u), ignoring", ID, SpellID);
-        return;
-    }
+    // For reload case
+    _trainers.clear();
 
-    if (GetTalentSpellCost(SpellID))
+    std::unordered_map<int32, std::vector<Trainer::Spell>> spellsByTrainer;
+    //if (QueryResult trainerSpellsResult = WorldDatabase.Query("SELECT ID, SpellID, MoneyCost, ReqSkillLine, ReqSkillRank, ReqAbility1, ReqAbility2, ReqAbility3, ReqLevel FROM trainer_spell"))
+    //if (QueryResult trainerSpellsResult = WorldDatabase.Query("SELECT b.ID, a.SpellID, a.MoneyCost, a.ReqSkillLine, a.ReqSkillRank, a.ReqLevel FROM trainer_spell AS a INNER JOIN trainer_spell AS b ON a.ID = -(b.SpellID) UNION SELECT * FROM trainer_spell WHERE SpellID > 0"))
+    if (QueryResult trainerSpellsResult = WorldDatabase.Query("SELECT b.ID, a.SpellID, a.MoneyCost, a.ReqSkillLine, a.ReqSkillRank, a.ReqAbility1, a.ReqAbility2, a.ReqAbility3, a.ReqLevel, a.VerifiedBuild FROM trainer_spell AS a "
+        "INNER JOIN trainer_spell AS b ON a.ID = -(b.SpellID) "
+        "UNION SELECT * FROM trainer_spell WHERE SpellID > 0"))
```

Comment:
Also: wouldn't it be better to first migrate all trainers to custom ids lists (basically, make everything refer to these negative spellids) and then do the same as master branch (id, not creatureId, linked by creature_default_trainer)? That would allow WPP to keep the same output code for both versions

Reply:
But how would WPP then know what to export the Id field as?

Reply:
WPP can autogenerate it with our standard `@ID` placeholder